desc/basis.py

```diff
@@ -1413,83 +1412,150 @@ def zernike_radial_poly(r, l, m, dr=0, exact="auto"):
         radial mode number(s)
-    m : ndarray of int, shape(K,)
+    m : ndarray of int, shape(K,) or integer
         azimuthal mode number(s)
     dr : int
         order of derivative (Default = 0)
 
     Returns
     -------
-    y : ndarray, shape(N,K)
+    out : ndarray, shape(N,K)
         basis function(s) evaluated at specified points
 
     """
-    m = jnp.abs(m)
-    alpha = m
-    beta = 0
-    n = (l - m) // 2
-    s = (-1) ** n
-    jacobi_arg = 1 - 2 * r**2
-    if dr == 0:
-        out = r**m * _jacobi(n, alpha, beta, jacobi_arg, 0)
-    elif dr == 1:
-        f = _jacobi(n, alpha, beta, jacobi_arg, 0)
-        df = _jacobi(n, alpha, beta, jacobi_arg, 1)
-        out = m * r ** jnp.maximum(m - 1, 0) * f - 4 * r ** (m + 1) * df
-    elif dr == 2:
-        f = _jacobi(n, alpha, beta, jacobi_arg, 0)
-        df = _jacobi(n, alpha, beta, jacobi_arg, 1)
-        d2f = _jacobi(n, alpha, beta, jacobi_arg, 2)
-        out = (
-            (m - 1) * m * r ** jnp.maximum(m - 2, 0) * f
-            - 4 * (2 * m + 1) * r**m * df
-            + 16 * r ** (m + 2) * d2f
-        )
-    elif dr == 3:
-        f = _jacobi(n, alpha, beta, jacobi_arg, 0)
-        df = _jacobi(n, alpha, beta, jacobi_arg, 1)
-        d2f = _jacobi(n, alpha, beta, jacobi_arg, 2)
-        d3f = _jacobi(n, alpha, beta, jacobi_arg, 3)
-        out = (
-            (m - 2) * (m - 1) * m * r ** jnp.maximum(m - 3, 0) * f
-            - 12 * m**2 * r ** jnp.maximum(m - 1, 0) * df
-            + 48 * (m + 1) * r ** (m + 1) * d2f
-            - 64 * r ** (m + 3) * d3f
-        )
-    elif dr == 4:
-        f = _jacobi(n, alpha, beta, jacobi_arg, 0)
-        df = _jacobi(n, alpha, beta, jacobi_arg, 1)
-        d2f = _jacobi(n, alpha, beta, jacobi_arg, 2)
-        d3f = _jacobi(n, alpha, beta, jacobi_arg, 3)
-        d4f = _jacobi(n, alpha, beta, jacobi_arg, 4)
-        out = (
-            (m - 3) * (m - 2) * (m - 1) * m * r ** jnp.maximum(m - 4, 0) * f
-            - 8 * m * (2 * m**2 - 3 * m + 1) * r ** jnp.maximum(m - 2, 0) * df
-            + 48 * (2 * m**2 + 2 * m + 1) * r**m * d2f
-            - 128 * (2 * m + 3) * r ** (m + 2) * d3f
-            + 256 * r ** (m + 4) * d4f
+    dr = jnp.asarray(dr).astype(int)
+
+    branches = [
+        _zernike_radial_vectorized,
+        _zernike_radial_vectorized_d1,
+        _zernike_radial_vectorized_d2,
+        _zernike_radial_vectorized_d3,
+        _zernike_radial_vectorized_d4,
+    ]
+    return switch(dr, branches, r, l, m, dr)
+
+
+@functools.partial(jnp.vectorize, excluded=(1, 2, 3), signature="()->(k)")
+def _zernike_radial_vectorized(r, l, m, dr):
+    """Calculation of Radial part of Zernike polynomials."""
+
+    def body_inner(N, args):
+        alpha, out, P_past = args
+        P_n2 = P_past[0]  # Jacobi at N-2
+        P_n1 = P_past[1]  # Jacobi at N-1
+        P_n = jacobi_poly_single(r_jacobi, N, alpha, 0, P_n1, P_n2)
+
+        # Only calculate the function at dr th index with input r
```

Comment:
what is dr th index?

Reply:
Ah my bad, at some point we had the switch function here, so it was to execute the intended derivative function

Reply:
So this is an old comment that is no longer needed?

Reply:
I changed this with the last commit